### PR TITLE
[READY] ✅  -- Introduce result type

### PR DESCRIPTION
**Pull request by mahmoudimus**
## Description of changes in release / Impact of release:

Given that Starlark does not support exceptions, there needs to be a way to map python exception handling and control flow to Starlark. One of the more popular ways to do this is to take inspiration from some of the functional programming constructs.

Copied this interface: https://github.com/MaT1g3R/option/blob/master/option/result.py

## Documentation

Documentation here: https://mat1g3r.github.io/option/

Use like this:

```python
load("@stdlib//jresult", _jresult="jresult")

f = _jresult.Result.success(1)
print(f)
print("-----")
print(f.map(lambda x: x+x).value())
print(f.is_ok == True)
print(f.is_err == False)
```

output:

```
0520 19:19:23.134 INFO: stdlib/larky.star:31:6: Result{value=1, error=null}
0520 19:19:23.134 INFO: stdlib/larky.star:32:6: -----
0520 19:19:23.139 INFO: stdlib/larky.star:33:6: 2
0520 19:19:23.140 INFO: stdlib/larky.star:34:6: True
0520 19:19:23.140 INFO: stdlib/larky.star:35:6: True
```

And:

```python
load("@stdlib//jresult", _jresult="jresult")

d = _jresult.Result.failure("oh no!")
print(d)
print(d.is_ok == False)
print(d.is_err == True)
```

output:

```
0520 19:19:23.143 INFO: stdlib/larky.star:37:6: Result{value=null, error=oh no!}
0520 19:19:23.144 INFO: stdlib/larky.star:38:6: True
0520 19:19:23.144 INFO: stdlib/larky.star:39:6: True
```

## Risks of this release
None.

### Is this a breaking change?
- [ ] Yes
- [X] No

### Is there a way to disable the change?
- [X] Use previous release
- [ ] Use a feature flag
- [ ] No

#### Additional details go here

Will make migration of python constructs using py2star much easier!

### CODE REVIEW

File: larky/src/main/java/com/verygood/security/larky/modules/types/results/Result.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+  }
+
+  @StarlarkMethod(name = "value")
+  public Object getValue() {
+    return value;
+  }
+
+  @StarlarkMethod(name = "error")
+  public EvalException getError() {
+    return error;
```
Comment (Agitolyev) [inline larky/src/main/java/com/verygood/security/larky/modules/types/results/Result.java:63]: As we provide accessors to underlying `Error` primitive, I think we should make an `Error` object as a legitimate first-class-citizen object of Larky.
 Which will allow doing sth like:
```
result  = doSomeFancyBusinessLogic()
err = result.getError()
if err.status() == 404:
    ...
elif err.status() == 401:
    ...
else:
    return Error(500)
```
I've intentionally used HTTP status codes, as they cover almost all use-cases (even IoT 418), familiar to everyone, and easy to interop.

**Comment by mahmoudimus** [comment]
@Agitolyev please take a look! Thank you!

File: larky/src/main/java/com/verygood/security/larky/modules/types/results/Result.java (hunk 1 above)
Comment (mahmoudimus) [reply larky/src/main/java/com/verygood/security/larky/modules/types/results/Result.java]: https://github.com/verygoodsecurity/starlarky/pull/121/commits/850b5cac5236a5c73cb3659b453020c72218d23a

**Review by Agitolyev — APPROVED** [review]

File: larky/src/main/java/com/verygood/security/larky/modules/types/results/Error.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+//  private Object value;
+//   private EvalException error;
+//
+//   Result(Object value, EvalException error) {
+//     this.value = value;
+//     this.error = error;
+//   }
+
+  private EvalException error;
```
[the remark this answers is not in the document]
Comment (mahmoudimus) [reply larky/src/main/java/com/verygood/security/larky/modules/types/results/Error.java:18]: Of course!

File: larky/src/main/java/com/verygood/security/larky/modules/types/results/Ok.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
+package com.verygood.security.larky.modules.types.results;
+
+import java.util.Objects;
+
+import net.starlark.java.eval.EvalException;
+import net.starlark.java.eval.Printer;
+
+public class Ok extends Result {
+
+  private Object value;
```
[the remark this answers is not in the document]
Comment (mahmoudimus) [reply larky/src/main/java/com/verygood/security/larky/modules/types/results/Ok.java]: https://github.com/verygoodsecurity/starlarky/pull/121/commits/23b7510d0b0cac174587cd18b95fdbada96e0f7a#diff-56e5f1d86cb68f8569311e981aec901cc9f85706dd1e515b12e19508244a24b4R12-R13

File: larky/src/main/java/com/verygood/security/larky/modules/types/results/Error.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mahmoudimus) [reply larky/src/main/java/com/verygood/security/larky/modules/types/results/Error.java]: https://github.com/verygoodsecurity/starlarky/pull/121/commits/23b7510d0b0cac174587cd18b95fdbada96e0f7a#diff-ad04ae9ad8be2d65f8dfa5f951a1345959293533329e166b036ab0249f1050b1R10

File: larky/src/main/java/com/verygood/security/larky/modules/ResultModule.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    extraKeywords = @Param(name = "kwargs", defaultValue = "{}"),
+    useStarlarkThread = true
+  )
+  public static Result safe(StarlarkCallable func, Tuple args, Dict<String, Object> kwargs, StarlarkThread thread) {
+    try {
+      return Result.ok(Starlark.call(thread, func, args, kwargs));
+    } catch (InterruptedException e) {
+      return Result.error(new EvalException(e.getMessage(), e));
+    } catch (EvalException e) {
+      return Result.error(e);
```
Comment (aslepakurov) [inline larky/src/main/java/com/verygood/security/larky/modules/ResultModule.java:62]: Might as well call static methods that are already here

```suggestion
      return ok(Starlark.call(thread, func, args, kwargs));
    } catch (InterruptedException e) {
      return error(new EvalException(e.getMessage(), e));
    } catch (EvalException e) {
      return error(e);
```

File: larky/src/main/java/com/verygood/security/larky/modules/types/results/Ok.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+  @Override
+  public Object getValue() {
+    return this.value;
+  }
+
+  @Override
+  public EvalException getError() {
+    return null;
+  }
```
Comment (aslepakurov) [inline larky/src/main/java/com/verygood/security/larky/modules/types/results/Ok.java:51]: Same as error

**Review by aslepakurov — APPROVED** [review]
Some comments

File: larky/src/main/java/com/verygood/security/larky/modules/types/results/Error.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+  @Override
+  public Object getValue() {
+    return null;
+  }
+
+  @Override
+  public EvalException getError() {
+    return this.exc;
+  }
```
[the remark this answers is not in the document]
Comment (mahmoudimus) [reply larky/src/main/java/com/verygood/security/larky/modules/types/results/Error.java:68]: It’s not clear how that works considering it’s doing an override of the Result interface. Some are expected to return null. Are you sure that will work ?

**Review by aslepakurov — APPROVED** [review]

**Pull request merged**